File: Dictionaries and Sets/2352. Equal Row and Column Pairs/2352. Equal Row and Column Pairs.py
class Solution:
    def equalPairs(self, grid: List[List[int]]) -> int:
        rows = defaultdict(int)

        for row in grid:
            rows[tuple(row)] += 1

        pairs = 0
        for j in range(len(grid)):
            key = []
            for i in range(len(grid)):
                key.append(grid[i][j])
            key = tuple(key)
            if key in rows:
                pairs += rows[key]

        return pairs

# Time: O(n^2)
# Space: O(n)

File: Dictionaries and Sets/2352. Equal Row and Column Pairs/2352. Equal Row and Column Pairs.js
/**
 * @param {number[][]} grid
 * @return {number}
 */
var equalPairs = function (grid) {
  const rows = {};

  for (const row of grid) {
    const r = row.join(",");
    rows[r] = (rows[r] || 0) + 1;
  }

  let pairs = 0;
  for (let j = 0; j < grid.length; j++) {
    let key = [];
    for (let i = 0; i < grid.length; i++) {
      key.push(grid[i][j]);
    }
    key = key.join(",");
    if (key in rows) {
      pairs += rows[key];
    }
  }

  return pairs;
};

// Time: O(n^2)
// Space: O(1)
